### PR TITLE
Only convert 1x1 conv to matmul if the width or height is not dynamic

### PR DESCRIPTION
**Pull request by rsuderman**
We cannot expand a single dynamic shape to multiple dynamic shapes. We should only
convert a 1x1 conv to a matmul if we can preserve the dynamic width / height.

### CODE REVIEW

**Comment by mariecwhite** [comment]
Nice find! Dynamic width/height can only work when the model consists of only convolutions (i.e. replace fully connected layers with 1x1 convs). Some background reading: https://arxiv.org/abs/1411.4038, and more digestible: https://towardsdatascience.com/implementing-a-fully-convolutional-network-fcn-in-tensorflow-2-3c46fb61de3b.

**Review by NatashaKnk — APPROVED** [review]

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/ConvertConv2D1x1ToMatmulPass.cpp [hunk 1]
```diff
@@ -27,17 +27,28 @@ class Convert1x1ConvolutionMatmulOp
...
+      return failure();
 
     auto inputShape = inputShapeType.getShape();
     auto filterShape = filterShapeType.getShape();
     auto outputShape = outputShapeType.getShape();
 
+    bool inputDynWidthHeight = inputShape[1] == ShapedType::kDynamicSize &&
+                               inputShape[2] == ShapedType::kDynamicSize;
+
+    // We cannot collapse if the width and heigh are dynamic.
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/ConvertConv2D1x1ToMatmulPass.cpp:49]: Not fully following everything here, but we can collapse when width and height are dynamic....

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/test/conv1x1_to_matmul.mlir [hunk 2]
```diff
@@ -18,3 +18,46 @@ func.func @conv_2d_1x1(%input: tensor<1x4x5x2xf32>, %filter: tensor<1x1x2x7xf32>
...
+    %0 = linalg.init_tensor [1, %d1, %d2, 7] : tensor<1x?x?x7xf32>
+    %1 = linalg.conv_2d_nhwc_hwcf {
+        dilations = dense<1> : tensor<2xi64>,
+        strides = dense<1> : tensor<2xi64>
+    } ins(%input, %filter : tensor<1x?x?x2xf32>, tensor<1x1x2x7xf32>) outs(%0 : tensor<1x?x?x7xf32>) -> tensor<1x?x?x7xf32>
+    return %1 : tensor<1x?x?x7xf32>
+}
+
+// CHECK: @conv_2d_1x1_fail_dyn
+// CHECK: linalg.conv_2d_nhwc_hwcf
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/test/conv1x1_to_matmul.mlir:63]: Nit: new line at end of file.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/ConvertConv2D1x1ToMatmulPass.cpp (hunk 1 above)
Comment (rsuderman) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/ConvertConv2D1x1ToMatmulPass.cpp]: Updated the comment. The problem is not the collapse but rather `Expand` does not support `? -> ?x?` to preserve both dynamic dimensions.

**Pull request merged**